dev/DynamicDependency/PackageDependency.h

```diff
@@ -0,0 +1,321 @@
+        m_lifetimeKind(other.m_lifetimeKind),
+        m_lifetimeArtifact(other.m_lifetimeArtifact),
+        m_options(other.m_options),
+        m_packageDependencyId(other.m_packageDependencyId)
+    {
+        User(other.m_user.get());
+    }
+
+    PackageDependency(PackageDependency&& other) :
+        m_user(std::move(other.m_user)),
+        m_packageFamilyName(std::move(other.m_packageFamilyName)),
+        m_minVersion(std::move(other.m_minVersion)),
+        m_packageDependencyProcessorArchitectures(std::move(other.m_packageDependencyProcessorArchitectures)),
+        m_lifetimeKind(std::move(other.m_lifetimeKind)),
+        m_lifetimeArtifact(std::move(other.m_lifetimeArtifact)),
+        m_options(std::move(other.m_options)),
+        m_packageDependencyId(std::move(other.m_packageDependencyId))
+    {
+    }
+
+    PackageDependency& operator=(const PackageDependency& other)
+    {
+        if (this != &other)
+        {
+            Reset();
+            User(other.m_user.get());
+            m_packageFamilyName = other.m_packageFamilyName;
+            m_minVersion = other.m_minVersion;
+            m_packageDependencyProcessorArchitectures = other.m_packageDependencyProcessorArchitectures;
+            m_lifetimeKind = other.m_lifetimeKind;
+            m_lifetimeArtifact = other.m_lifetimeArtifact;
+            m_options = other.m_options;
+            m_packageDependencyId = other.m_packageDependencyId;
+        }
+        return *this;
+    }
+
+    PackageDependency& operator=(PackageDependency&& other)
+    {
+        if (this != &other)
+        {
+            Reset();
+            m_user = std::move(other.m_user);
+            m_packageFamilyName = std::move(other.m_packageFamilyName);
+            m_minVersion = std::move(other.m_minVersion);
+            m_packageDependencyProcessorArchitectures = std::move(other.m_packageDependencyProcessorArchitectures);
+            m_lifetimeKind = std::move(other.m_lifetimeKind);
+            m_lifetimeArtifact = std::move(other.m_lifetimeArtifact);
+            m_options = std::move(other.m_options);
+            m_packageDependencyId = std::move(other.m_packageDependencyId);
+        }
+        return *this;
+    }
+
+    void Reset()
+    {
+        m_user.reset();
+        m_packageFamilyName.clear();
+        m_minVersion = PACKAGE_VERSION{};
+        m_packageDependencyProcessorArchitectures = MddPackageDependencyProcessorArchitectures::None;
+        m_lifetimeKind = MddPackageDependencyLifetimeKind::Process;
+        m_lifetimeArtifact.clear();
+        m_options = MddCreatePackageDependencyOptions::None;
+        m_packageDependencyId.clear();
+    }
+
+    void GenerateId();
+
+    void User(PSID user)
+    {
+        if (user)
+        {
+            const auto sidLength{ GetLengthSid(user) };
+            wil::unique_any_psid sid{ static_cast<PSID>(::LocalAlloc(LMEM_FIXED, sidLength)) };
```

Comment:
why not HeapAlloc?
See note at: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-localalloc

Reply:
`wil::unique_any_psid` uses `LocalFree` to free the SID

Reply:
And PSID is weird -- PSID!=SID*, APIs returning PSID typically allocate it via LocalAlloc and WIL's unique_any_psid works exactly as expected as a RAII PSID (and is the only WIL smart pointer that truly acts like a PSID)